Add UseUsingScopeModifierInNewRunspaces rule

Jawz84 opened the pull request:
## PR Summary

Fixes #1410 

- [x] write tests
- [x] write documentation
- [x] draft rudimentary rule
- [x] tune rule until all tests for this rule pass (except the test for built-in vars)
- [x] adapt other rules that start failing
- [x] add support for built-in variable detection
- [x] expand functionality to other commands /situations where this check makes sense
  - [x] added support for InlineScript, Start-(Thread)Job and Invoke-Command
- [x] finetune
- [x] implement suggested correction (Replace $var with $using:var)
  - [x] add test for suggested correction
- [x] refactor to AstVisitor/AstVisitor2
- [x] add support for DSC Script Resource (SetScript, GetScript, TestScript)

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] Make sure you've added a new test if existing tests do not effectively test the code changed and/or updated documentation
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.

bergmeister commented on Tests/Rules/AvoidUnInitializedVarsInNewRunspaces.tests.ps1:
```diff
@@ -0,0 +1,90 @@
+$directory = Split-Path -Parent $MyInvocation.MyCommand.Path
+$testRootDirectory = Split-Path -Parent $directory
+
+Import-Module (Join-Path $testRootDirectory "PSScriptAnalyzerTestHelper.psm1")
```
I don't think we use the test helper (or do we?) and therefore I think we can remove the first 4 lines.

Jawz84 replied:
Good point, we don't, so that can indeed be removed.

bergmeister commented on Tests/Rules/AvoidUnInitializedVarsInNewRunspaces.tests.ps1:
```diff
@@ -0,0 +1,90 @@
+$ruleName = "PSAvoidUnInitializedVarsInNewRunspaces"
+
+$settings = @{
+    IncludeRules = @($ruleName)
```
I know some old tests do this but since `ruleName` is not re-used a 2nd time, I would not declare a variable for it and just inline it

Jawz84 replied:
Apparently I peeked at an old test for template :-) will inline it.

bergmeister commented on Tests/Rules/AvoidUnInitializedVarsInNewRunspaces.tests.ps1:
```diff
@@ -0,0 +1,90 @@
+    IncludeRules = @($ruleName)
+}
+
+Describe "AvoidUnInitializedVarsInNewRunspaces" {
```
Generally: Good test coverage and clean code. That's the right level we look for. At the end we might come back and add a few special test cases but looks great so far.

bergmeister commented on Rules/AvoidUnInitializedVarsInNewRunspaces.cs:
```diff
@@ -0,0 +1,173 @@
+        /// <returns>An enumerable containing diagnostic records</returns>
+        private IEnumerable<DiagnosticRecord> AnalyzeScriptBlockAst(ScriptBlockAst scriptBlockAst, string fileName)
+        {
+            var foreachObjectCmdletNamesAndAliases = Helper.Instance.CmdletNameAndAliases("Foreach-Object");
```
the helper instance is using a cache that has queried the actual system that it is running on, therefore behaviour will depend on the system (which can be good and bad) and has a bit of a 'cold start' performance impact because it uses `get-command` behind the scenes for that. If the rule was just for `ForEach-Object` then I'd say let's just hard-code it since the aliases are the same across every ps version and platform. But if you intend to increase the scope of the rule to other command like `invoke-command`, etc. then it's maybe better to leave it like it is at the moment. Therefore consider this just as a remark for the moment, let's see what Rob thinks about it.

bergmeister commented on Rules/AvoidUnInitializedVarsInNewRunspaces.cs:
```diff
@@ -0,0 +1,173 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
I haven't read the logic in detail yet but from a high level it looks ok, you are using the right methods and concepts to search through the AST, so feel free to continue

bergmeister commented:
You are on the right track.  I suggest to continue focusing on the code of the rule itself. Don't worry about the rest like docs or some other test failures at the moment, they will be easy to fix at the end.

Jawz84 commented:
@bergmeister I've hit a milestone. At this point the code seems to be working as intended. I have added one minor feature which may need to be in a separate PR: I've added the option to use `Test-ScriptAnalyzer` with a `-RuleToTest` parameter, so I could easily make adjustments, run `.\build.ps1`, and run `Test-ScriptAnalyzer -RuleToTest 'avoiduni' -ShowAll` to check if the change made it worse or better :-).

Can we have a little brainstorm on where this rule may be applied to, other than just for `Foreach-Object -Parallel`? I'm thinking of maybe `Invoke-Command`, but other than that?

Would you like to start a review round at this point, or better after we have a clear picture of how to move forward?

Jawz84 commented:
The failure in Ubuntu 1804 build is the same Authenticode glitch we saw earlier.

Jawz84 commented:
// brainstorm

Other situations where this rule can be applied:

- `Start-Job/Start-ThreadJob -scriptBlock {$using:foo}`
                Here the catch is, we need to be sure we check the right scriptblock. The rule does not apply to the `-InitializationScript` scriptblock parameter.
            
- `Invoke-Command -ComputerName "baz" -ScriptBlock {$using:foo}`
                The catch here, is that the `-ComputerName` parameter _has_ to be used, otherwise `$using` is prohibited. 
                Also tricky; one can open a persistent session, and use subsequent `invoke-command` calls to that, where variables previous calls are still valid (more info [here](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_remote_variables?view=powershell-7#using-remote-variables))
                
                ```powershell
                $session = new-pssession -computername "baz"
                Invoke-Command -session $session -scriptblock {$foo = "foo" }
                Invoke-Command -session $session -scriptblock {Write-Output $foo} # works
                ```
                
- `workflow {$foo = 'foo'; inlinescript {$using:foo} }`
                On Windows PowerShell only. (see get-help about_InlineScript)

Jawz84 commented:
@mklement0 thanks for creating the powershell-docs PR for documenting this. That gave me the quick overview I needed. If you have any suggestions, I'd love to hear them, thanks.

mklement0 commented:
Thanks, @Jawz84, a great rule to add, and it would be great if it covered the other `$using:` contexts too (as you outline above).

One pending PR to watch out for: https://github.com/PowerShell/PowerShell/pull/11829, which is trying to eliminate overzealous  `$using:` checks for _nested_ script blocks, as described in https://github.com/PowerShell/PowerShell/issues/11817

As for the terminology in the rule documentation: [`about_Remote_Variables`](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_Remote_Variables) refers to `$using:` as the  `Using` _scope modifier_ - it's not a _directive_.

Jawz84 commented:
Thanks for pointing these out @mklement0 !

Jawz84 commented:
The rule is working now. I am not completely satisfied with what the implementation looks like, and I'm likely to do some cleanup and refactor. If @mklement0 or @bergmeister have insights how to best convert this to cleaner code, I'm all ears. 

Second question: when starting out, I named the rule `AvoidUnInitializedVarsInNewRunspaces` Is that the best, most descriptive name for this rule?

Jawz84 commented:
I had a chat with @chrislgardner discussing the name of the rule and the warning message. He suggested we could change them to:
- rulename: PSUseUsingInNewRunspaces
- warning msg: The variable '{0}' is not declared in this ScriptBlock and should have the $using: scope modifier if declared outside of it.
I like both suggestions. Any other Ideas? @felixfbecker

bergmeister commented on build.psm1:
```diff
@@ -289,7 +289,14 @@ function Start-ScriptAnalyzerBuild
+
+        [switch]$ShowAll,
+
+        [string]$RuleToTest
```
Amazing, I was thinking of writing something similar a while ago but never came round to actually doing it. But since there are also engine tests, I think just the path to the pester tests file would be better

Jawz84 replied:
I have some ideas for this, including an argument completer. Let's flesh this out further in a separate PR. I'll revert this for this PR, and submit another one for improving the tools.

Jawz84 commented:
Thanks for the review! Going to look into in detail later. Great points from what I've seen so far. This will make it better!

Jawz84 commented on Rules/AvoidUnInitializedVarsInNewRunspaces.cs:
```diff
@@ -0,0 +1,363 @@
+                        .Text
+                        .Trim();
+                }
+                catch
```
My reasoning was, that if -Session is used without an argument (and the runtime would allow it), it would be OK to just catch and continue. It a broken situation anyways. 
What would be a better way to convey this? Maybe like this:
```csharp
if (commandAst.CommandElements.Count > commandAst.CommandElements.IndexOf(sessionParameterAst)) {
    sessionName = commandAst
                        .CommandElements[
                            commandAst
                                .CommandElements
                                .IndexOf(sessionParameterAst) + 1]
                        .Extent
                        .Text
                        .Trim();
}
```

Jawz84 replied:
I've refactored it a bit.

Jawz84 commented on Rules/AvoidUnInitializedVarsInNewRunspaces.cs:
```diff
@@ -0,0 +1,363 @@
+                {
+                    sessionDictionary.Add(sessionName, new List<Ast>());
+                    sessionDictionary[sessionName].Add(ast);
+                }
```
Thank you. During cleanup, I looked at this part twice because it smelled, but couldn't see it. This is what it needs to be, cheers.

Jawz84 commented on RuleDocumentation/AvoidUnInitializedVarsInNewRunspaces.md:
```diff
@@ -0,0 +1,27 @@
+
+## How to Fix
+
+Within `Foreach-Object -Parallel {}`, instead of just using a variable from the parent scope, you have to use the `using:` directive:
```
Great point, added a list of applicable commands, and an example for each one. I found out I had the 'directive' thing here still. Changed that too.

Jawz84 commented on Tests/Rules/AvoidUnInitializedVarsInNewRunspaces.tests.ps1:
```diff
@@ -0,0 +1,197 @@
+$settings = @{
```
thanks 👍

bergmeister commented:
@rjmholt Are you happy with this as well on a high level as well?
From a risk perspective: The new rule runs by default but its low level code paths get invoked only in a small subset of scenarios. Therefore, although there are a few index accessors, the risk is relatively small that they could index into a non-existing index in some edge cases and therefore I don't think we have to apply a fully defensive coding approach.

Jawz84 commented:
@bergmeister I found time to add an implementation for suggested correction ('Replace $var with $using:var'). I kept it as lean as possible.

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,375 @@
+            var astsToProcess = new List<Ast>();
+            var nonAssignedNonUsingVars = new List<VariableExpressionAst>();
+
+            astsToProcess.AddRange(
```
You'd probably be much better off writing an `AstVisitor2` to accomplish this, since that allows you to simply override some methods to visit the ASTs you're looking for.

[Here](https://github.com/PowerShell/PSScriptAnalyzer/blob/f389aa42166a8fdbf293432eb0c8f42aa7c11508/Rules/CompatibilityRules/UseCompatibleSyntax.cs#L152-L376)'s an example

bergmeister replied:
do you know in which PS version AstVisitor2 was introduced as PSSA still has to support PSv3 unfortunately and therefore we might need conditional compilation for that

Jawz84 replied:
on it

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,375 @@
+                FindAllStartJobAsts(scriptBlockAst));
+
+            astsToProcess.AddRange(
+                FindAllForeachParallelAsts(scriptBlockAst));
```
It also looks like we're searching over all the commands in a script block for each individual command type.

For a script with thousands of lines, that's probably not ideal (although still fairly quick I imagine)

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,375 @@
+                ProcessInvokeCommandSessionAsts(scriptBlockAst));
+
+
+            foreach (var ast in astsToProcess)
```
```suggestion
            foreach (Ast ast in astsToProcess)

```

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,375 @@
+            foreach (var ast in astsToProcess)
+            {
+                if (ast == null)
+                    continue;
```
Please use braces for conditional statements

Jawz84 replied:
I have fixed this throughout, thanks

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,375 @@
+                    ruleName: GetName(),
+                    severity: DiagnosticSeverity.Warning,
+                    scriptPath: fileName,
+                    ruleId: variableExpression.ToString(),
```
Is this the correct rule ID? I thought rule ID was supposed to be the name of the rule generating the diagnostic

Jawz84 replied:
No, you are right, it is not. I have changed it to the rule name.

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,375 @@
+
+        private static IEnumerable<CorrectionExtent> GetSuggestedCorrections(Ast ast)
+        {
+            var varWithUsing = "$using:" + ast.Extent.Text.TrimStart('$');
```
Rather than use `ast.Extent.Text` you should use `ast.VariablePath.UserPath`:

```suggestion
        private static IEnumerable<CorrectionExtent> GetSuggestedCorrections(VariableExpressionAst ast)
        {
            var varWithUsing = $"$using:{ast.VariablePath.UserPath}";

```

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,375 @@
+                )
+            };
+
+            return corrections;
```
```suggestion
            return new CorrectionExtent[]
            {
                new CorrectionExtent(
                    startLineNumber: ast.Extent.StartLineNumber,
                    endLineNumber: ast.Extent.EndLineNumber,
                    startColumnNumber: ast.Extent.StartColumnNumber,
                    endColumnNumber: ast.Extent.EndColumnNumber,
                    text: varWithUsing,
                    file: ast.Extent.File,
                    description: description
                )
            };
```

Jawz84 replied:
thanks

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,375 @@
+
+        private static IEnumerable<VariableExpressionAst> ProcessInvokeCommandSessionAsts(ScriptBlockAst scriptBlockAst)
+        {
+            var invokeCommandCmdletNamesAndAliases = Helper.Instance.CmdletNameAndAliases("Invoke-Command");
```
```suggestion
            IEnumerable<string> invokeCommandCmdletNamesAndAliases = Helper.Instance.CmdletNameAndAliases("Invoke-Command");

```

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,375 @@
+            var result = new List<VariableExpressionAst>();
+
+            // The shortest unambiguous name for parameter -Session is 'session' (SessionName and SessionOption) also exist.
+            var scriptBlockExpressionAstsToAnalyze = scriptBlockAst.FindAll(
```
```suggestion
            IEnumerable<Ast> scriptBlockExpressionAstsToAnalyze = scriptBlockAst.FindAll(

```

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,375 @@
+
+            // The shortest unambiguous name for parameter -Session is 'session' (SessionName and SessionOption) also exist.
+            var scriptBlockExpressionAstsToAnalyze = scriptBlockAst.FindAll(
+                predicate: a => a is ScriptBlockExpressionAst scriptBlockExpressionAst &&
```
This is too complex to be an inline expression I think. Better to move it to a method

rjmholt replied:
This is really suited to a visitor though

Jawz84 replied:
Still working on this one.

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,375 @@
+                predicate: a => a is ScriptBlockExpressionAst scriptBlockExpressionAst &&
+                                scriptBlockExpressionAst.Parent is CommandAst commandAst &&
+                                invokeCommandCmdletNamesAndAliases
+                                    .Contains(commandAst.GetCommandName(), StringComparer.OrdinalIgnoreCase) &&
```
`commandAst.GetCommandName()` can be null, which will cause a crash here

rjmholt replied:
Example:

```powershell
{ & $commandName }.Ast.EndBlock.Statements[0].PipelineElements[0].GetCommandName()
```

Jawz84 replied:
Thanks, noted, I'm thinking of adding a test along these lines to be sure.

Jawz84 commented:
> Sorry I've taken a while to review. Been working on the VSCode extension for the PS 7 release.

No problem at all! Obviously VSCode takes precedence over this. I'm glad to get reviews from you all.

> * For this to be a default rule I think the bar is high and we should ensure that the rule is implemented in a maintainable and performant way. I think an important change to this rule would be to either migrate it to an `AstVisitor2` implementation or to at least collect the ASTs in one swoop with predicates defined in methods. For PSES to run this every second currently, I'm worried the overhead of the allocations of all the lists and delegates will hurt us.

Reading up on AstVisitor2, thank you for the example link. 

> * How does this interact with `PSUseDeclaredVariablesMoreThanAssignment`? Will this generate double diagnostics? If so, we will get user reports complaining about that

So you mean would there be possible situations where both `PSUseDeclaredVariablesMoreThanAssignment` and `PSUseUsingScopeModifierInNewRunspaces`  would apply to the same variable. I don't know. What would a double diagnostic for the same script extent do, for example in VSCode? It would be confusing for the user at least, but would it generate errors too? What would be the best way to prevent this. There is no way of 'knowing' from within a rule if another rule already applies to that scriptextent is there?

Jawz84 commented:
https://twitter.com/brwilkinson/status/1236006065407328256?s=19

Jawz84 commented:
I have come a long way refactoring to AstVisitor/AstVisitor2. I have not yet finished the refactor for the part of the code that analyzes `Invoke-Command -Session`. I'm working on that.

Jawz84 commented:
Refactor to AstVisitor is done. Tests fail on build for Windows Powershell, there is obviously something I did wrong with the compiler directives.

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,413 @@
+#if !(PSV3 || PSV4)
+    internal class SyntaxCompatibilityVisitor : AstVisitor2
+#else
+        private class SyntaxCompatibilityVisitor : AstVisitor
```
You want this to be an `internal` rather than a `private` class. Alternatively you can move it inside `UseUsingScopeModifierInNewRunspaces` and make it private. But this is why PS3 compile is failing

Jawz84 replied:
That was fast, thank you! Came here to say I found exactly this :-)

Jawz84 replied:
Build now fails on 
```
Executing script C:\Users\Jos.Koelewijn\github\PSScriptAnalyzer\Tests\Rules\UseCompatibleCommands.Tests.ps1

  Describing UseCompatibleCommands

    Context Targeting a single profile

    Context Checking a file against many targets

    Context Checking PSSA repository scripts
      [-] Ensures that PSSA build scripts are cross compatible with everything 352ms
        Expected 0, because PSUseCompatibleCommands, PSUseCompatibleCommands, but got 2.
        336:             $diagnostics.Count | Should -Be 0 -Because ($diagnostics.RuleName -join ', ')
        at <ScriptBlock>, C:\Users\Jos.Koelewijn\github\PSScriptAnalyzer\Tests\Rules\UseCompatibleCommands.Tests.ps1: line 336
```

Apparently something I did affected how `PSUseCompatibleCommands` and `PSUseCompatibleCommands` are evaluated here.

Jawz84 replied:
I'll have to see about this later.

rjmholt commented on build.psm1:
```diff
@@ -704,3 +714,12 @@ function Copy-CrossCompatibilityModule
+Function RuleNameCompleter {
+    param($commandName, $parameterName, $wordToComplete, $commandAst, $fakeBoundParmeter)
+
+    Get-Childitem -Path "${projectRoot}\Tests\Rules" -Filter "*$wordToComplete*.tests.ps1" -Name |
```
Might be here where the compatibility issue is occurring. And does the `-Name` parameter need an argument?

Jawz84 replied:
This whole change should not have ended up here. I accidentally pushed a few files that did not belong. Reverted, and all good now.

Jawz84 commented:
@bergmeister @rjmholt could you have another look? Looking forward to your review, whenever you have time.

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,430 @@
+                    commandAst.CommandElements.IndexOf(scriptBlockExpressionAst) - 1] as CommandParameterAst;
+
+            if (IsInlineScriptBlock(cmdName) || 
+                IsJobScriptBlock(cmdName, scriptBlockParameterAst) ||
```
This looks for an explicit parameter name, but `Start-Job { <# do stuff #> }` seems quite common. Is it worth looking for explicit parameter?

Jawz84 replied:
I think it is, yes. I see the comment explaining the why of this was lost in the refactor. For `Start-Job`, there can be a situation where there are _two_ ScriptPlocks passed in as parameters. The other one being `-InitializationScript`, which does not support the `using` scope modifier. That is why we need this information. I can add in a comment in the code for this. 
Did I interpret your comment the right way?

rjmholt replied:
Ah! Yep I see now. Thanks for the explanation

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,430 @@
+
+            if (IsInvokeCommandSessionScriptBlock(cmdName, commandAst))
+            {
+                var sessionName = GetSessionName(commandAst);
```
The type of the LHS isn't indicated on the RHS here. Would it be possible to use an explicit type signature? (Here and with other `var` uses resulting from a method call)

Jawz84 replied:
Yes, no problem, done. Maybe there is a setting I can change to make Visual Studio stop complaining. Visual Studio keeps pointing out to me that I can change these to `var`, but I certainly see your point for readability.

rjmholt replied:
Yeah VS seems to have some configurations that are overzealous. From my perspective, I spend a lot of time reading code, and often not in an IDE. That means being able to infer the type of something on the same line it's defined is valuable, so my general rule is that `var` should be used only when the type is expressed on the right hand side:

- `var x = new Thing();`
- `var x = y as Thing;`
- `var x = JsonConvert.Convert<Thing>(str);`
- The iffy `var x = Thing.Create();`

But the antipatterns in my mind are things like:

```csharp
var x = w.Method();
var y = x.Property;
var z = y.AnotherMethod(); // If we got x wrong, the compile error might only show here
```

or

```csharp
var x = 7; // int has the same number of characters as var...
```

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,430 @@
+        private static IEnumerable<VariableExpressionAst> FindVarsInAssignmentAsts(Ast ast)
+        {
+            // Find all variables that are assigned within this ast
+            return ast.FindAll(
```
Converting away from this is really what I meant by converting to an AstVisitor; implementing methods like `VisitAssignmentStatement`.

But I won't block on it.

Instead, I think you should do the following:

- Convert the predicate to a static method
- Look for assignment ASTs rather than variable ASTs
- Use a `foreach` loop rather than LINQ

Jawz84 replied:
I'm going to look into it for both `FindVarsInAssignmentAsts` and `FindNonAssignedNonUsingVarAsts`. I'm having a hard time wrapping my head around this, but will try.

rjmholt replied:
It's stateful and easier to get wrong. I think keeping the `FindAll` method is fine, you just want to turn your predicates into static methods, so they don't have to allocate closures

Jawz84 replied:
I have given this some more thought. I feel if I use a visitor for this, I have to keep track of scope for each of the variables I encounter, which will become complicated and error prone real quick. I think I'll go the easier route then, and keep the `FindAll` but swap out the predicate for a static method as you explained (and invert the search to assignments).

rjmholt replied:
Yeah I agree

Jawz84 replied:
I think I got this one, see latest commit.

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,430 @@
+            Ast ast, IEnumerable<VariableExpressionAst> varsInAssignments)
+        {
+            // Find all variables that are not locally assigned, and don't have $using: scope modifier
+            return ast.FindAll(
```
Ditto above

Jawz84 replied:
see above.

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,430 @@
+                                    varExpr.Parent is AssignmentStatementAst assignment &&
+                                    assignment
+                                        .Left
+                                        .Equals(varExpr),
```
This can be bypassed with something like:

```
[int]$x = 10
```

The assignment is valid, but the LHS of the assignment doesn't equal the variable AST

Jawz84 replied:
Great catch! You are absolutely right. I changed two tests to make sure we keep checking for this. I'll have to take this into account when refactoring to visitor method.

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,430 @@
+                                        .Instance
+                                        .HasSpecialVars(varAst.VariablePath.UserPath),
+                    searchNestedScriptBlocks: true)
+                .Select(a => a as VariableExpressionAst).ToList();
```
Here you should definitely use a `foreach` and build the list directly to prevent double allocation

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,430 @@
+        /// <param name="commandAst"></param>
+        private static string GetSessionName(CommandAst commandAst)
+        {
+            if (!(commandAst.CommandElements.First<Ast>(
```
An index here would be better for performance

Jawz84 replied:
Can you help me learn what you mean by giving an example please? I fail to see how to do this rewrite.

rjmholt replied:
Sure! I see that really `First` takes a predicate, so we can't simply index it, but something like this:

```csharp
private static bool TryGetSessionNameFromInvokeCommand(CommandAst invokeCommandAst, out string sessionName)
{
    // Sift through Invoke-Command parameters to find the value of the -Session parameter
    // Start at 1 to skip the command name
    for (int i = 1; i < invokeCommandAst.CommandElements.Count; i++)
    {
        // We need a parameter
        if (!(invokeCommandAst.CommandElements[i] is CommandParameterAst parameterAst))
        {
            continue;
        }

        // The parameter must be called "Session"
        if (!parameterAst.ParameterName.Equals("Session", StringComparison.OrdinalIgnoreCase))
        {
            continue;
        }

        // If we have a partial AST, ensure we don't crash
        if (i + 1 >= invokeCommandAst.CommandElements.Count)
        {
            break;
        }

        // Session could be dynamically specified
        // We can only deal with string constants
        if (!(invokeCommandAst.CommandElements[i + 1] is StringConstantExpressionAst stringAst))
        {
            break;
        }

        sessionName = stringAst.Value;
        return true;
    }

    sessionName = null;
    return false;
}
```

Jawz84 replied:
Thanks for this. Almost there now, tbc

Jawz84 replied:
Done. I have changed the last test, because `-Session` expects a Session object, usually supplied by a variable. The name of the variable is now the entry we will use to record what belongs to which session.

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,430 @@
+                return "";
+            }
+
+            var sessionParamAstIndex = commandAst.CommandElements.IndexOf(sessionParameterAst);
```
```suggestion
            int sessionParamAstIndex = commandAst.CommandElements.IndexOf(sessionParameterAst);

```

Jawz84 replied:
fixed in different commit.

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,430 @@
+        private void AnalyzeScriptBlock(ScriptBlockExpressionAst scriptBlockExpressionAst)
+        {
+            var nonAssignedNonUsingVarAsts = FindNonAssignedNonUsingVarAsts(scriptBlockExpressionAst,
+                FindVarsInAssignmentAsts(scriptBlockExpressionAst)).ToList();
```
The method below doesn't depend on the argument being a list, so this just double allocates and can be removed

```suggestion
                FindVarsInAssignmentAsts(scriptBlockExpressionAst));

```

Jawz84 commented on Tests/Rules/UseUsingScopeModifierInNewRunspaces.tests.ps1:
```diff
@@ -0,0 +1,280 @@
+            }
+    )
+
+    It "should emit for: <Description>" -TestCases $testCases {
```
good catch, fixed. thanks.

Jawz84 commented on Tests/Rules/UseUsingScopeModifierInNewRunspaces.tests.ps1:
```diff
@@ -0,0 +1,280 @@
+}
+
+Context "Should not detect anything" {
+    $testCases = @(
```
Moved testcases to `BeforeAll` blocks.

Jawz84 commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,430 @@
+    }
+
+#if !(PSV3 || PSV4)
+    internal class SyntaxCompatibilityVisitor : AstVisitor2
```
The rule itself is public, see line 22: `public class UseUsingScopeModifierInNewRunspaces : IScriptRule`. This class is only the visitor helper class. I don't think this class has much use outside of the scope of this rule. Would it be clearer to make this a private nested class?

Jawz84 replied:
I think it is clearer if I turn the visitor into a private class within the rule class.

Jawz84 commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,430 @@
+    {
+        private const DiagnosticSeverity Severity = DiagnosticSeverity.Warning;
+
+        private static readonly string[] s_dscScriptResourceCommandNames = {"GetScript", "TestScript", "SetScript"};
```
Commit accepted. I have not seen this before. Can you explain the benefit to me? I'd love to learn, thanks.

Jawz84 commented:
@rjmholt I've been thinking about what to optimize, and especially: how to measure if it has worked. When I run tests, most are around 15ms, but some stand out and take longer: 
![image](https://user-images.githubusercontent.com/26464885/76690540-3cf7c680-6641-11ea-9c25-cc3cabfc1204.png)
I have tried Visual Studios profiling tool and the diagnostic tools, but none the wiser really. How do I visualise which codepaths make the `Non-DSC function with the name SetScript {}` test take _156ms_ instead of the more typical 15ms?

Jawz84 commented:
@rjmholt did you have a chance to look at this yet? I'm starting to fear we might be accidentally waiting _for each other_

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,493 @@
+            }
+
+            /// <summary>
+            /// VarsInAssignments: helper function to prevent allocation of closures for FindAll predicate.
```
I know it's the convention in the rule comments, but there's no need to repeat the method name in the comment. The comment should add information that isn't present in the name or type signature of the thing it's attached to.

rjmholt replied:
No need to change it

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,493 @@
+            /// VarsInAssignments: helper function to prevent allocation of closures for FindAll predicate.
+            /// </summary>
+            /// <param name="ast"></param>
+            private static bool VarsInAssignments(Ast ast)
```
A predicate method should have a boolean name that starts with a verb, like `IsAssignmentStatement`

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,493 @@
+            /// NonUsingNonSpecialVariables: helper function to prevent allocation of closures for FindAll predicate.
+            /// </summary>
+            /// <param name="ast"></param>
+            private static bool NonUsingNonSpecialVariables(Ast ast)
```
```suggestion
            private static bool IsNonUsingNonSpecialVariableAst(Ast ast)

```

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,493 @@
+            private void AnalyzeScriptBlock(ScriptBlockExpressionAst scriptBlockExpressionAst)
+            {
+                    IEnumerable<VariableExpressionAst> nonAssignedNonUsingVarAsts = FindNonAssignedNonUsingVarAsts(scriptBlockExpressionAst,
+                    FindVarsInAssignmentAsts(scriptBlockExpressionAst));
```
```suggestion
                    IEnumerable<VariableExpressionAst> nonAssignedNonUsingVarAsts = FindNonAssignedNonUsingVarAsts(scriptBlockExpressionAst,
                    FindVarsInAssignmentAsts(scriptBlockExpressionAst));
```suggestion
                IEnumerable<VariableExpressionAst> nonAssignedNonUsingVarAsts = FindNonAssignedNonUsingVarAsts(scriptBlockExpressionAst,
                FindVarsInAssignmentAsts(scriptBlockExpressionAst));

```
```

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,493 @@
+            /// Example: `$foo = "foo"` ==> the VariableExpressionAst for $foo is returned
+            /// </summary>
+            /// <param name="ast"></param>
+            private static IEnumerable<VariableExpressionAst> FindVarsInAssignmentAsts(Ast ast)
```
Rather than return an `IEnumerable<VariableExpressionAst>`, this could perhaps return an `IReadOnlyDictionary<string, VariableExpressionAst>`, a case-insensitive dictionary of variables by variable name, allowing the lookup later to be constant time rather than linear

Jawz84 replied:
See comment for line 251. I like it!

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,493 @@
+                    {
+                        yield return variable;
+                    }
+                    else if (assignment.Left is ConvertExpressionAst conversion)
```
To get around the issue you basically need something like this:

```csharp
private static VariableExpressionAst GetVariableFromExpression(ExpressionAst expression)
{
    switch (expression)
    {
        case VariableExpressionAst varAst:
            return varAst;

        case ConvertExpressionAst convertAst:
            return GetVariableFromExpression(convertAst.Child);

        case AttributedExpressionAst attributedAst:
            return GetVariableFromExpression(attributedAst.Child);

        default:
            // return null or throw depending on what you need
            // perhaps this should be TryGetVariableFromExpression(out VariableExpressionAst varAst)
            // and return false here
    }
}
```

Jawz84 replied:
Thanks, that is a beautiful way of solving this rabbit hole I didn't see.

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,493 @@
+                {
+                    var variable = varAst as VariableExpressionAst;
+
+                    foreach (var expressionAst in varsInAssignments)
```
This would be where you do a dictionary lookup rather than the `foreach`

Jawz84 replied:
I like this idea a lot! Working on this, it affects a few other places too.

Jawz84 replied:
done now.

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,493 @@
+                Ast ast, IEnumerable<VariableExpressionAst> varsInAssignments)
+            {
+                // Find all variables that are not locally assigned, and don't have $using: scope modifier
+                foreach (Ast varAst in ast.FindAll(NonUsingNonSpecialVariables, true))
```
The foreach will do the conversion for you, so you don't need to specify `Ast` here and have `as VariableExpressionAst` below. Just put `VariableExpressionAst` here

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,493 @@
+            /// <param name="ast"></param>
+            private static IEnumerable<CorrectionExtent> GetSuggestedCorrections(VariableExpressionAst ast)
+            {
+                var varWithUsing = $"$using:{ast.VariablePath.UserPath}";
```
```suggestion
                string varWithUsing = $"$using:{ast.VariablePath.UserPath}";

```

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,493 @@
+            /// <param name="nonAssignedNonUsingVarAsts"></param>
+            private void GenerateDiagnosticRecords(IEnumerable<VariableExpressionAst> nonAssignedNonUsingVarAsts)
+            {
+                foreach (var variableExpression in nonAssignedNonUsingVarAsts)
```
```suggestion
                foreach (VariableExpressionAst variableExpression in nonAssignedNonUsingVarAsts)

```

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,493 @@
+            {
+                foreach (var variableExpression in nonAssignedNonUsingVarAsts)
+                {
+                    if (variableExpression == null)
```
This check shouldn't be needed if the caller is performing as expected. A null check makes sense with a public API, but with a private one it will just disguise a possible bug

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,493 @@
+                        severity: Severity,
+                        scriptPath: _analyzedFilePath,
+                        ruleId: _rule.GetName(),
+                        suggestedCorrections: GetSuggestedCorrections(ast: variableExpression)));
```
```suggestion
                    string diagnosticMessage = string.Format(
                        CultureInfo.CurrentCulture,
                        Strings.UseUsingScopeModifierInNewRunspacesError,
                        variableExpression.ToString());

                    _diagnosticAccumulator.Add(
                        new DiagnosticRecord(
                            diagnosticMessage,
                            variableExpression.Extent,
                            _rule.GetName(),
                            Severity,
                            _analyzedFilePath,
                            ruleId: _rule.GetName(),
                            GetSuggestedCorrections(ast: variableExpression)));

```

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,493 @@
+                return _invokeCommandCmdletNamesAndAliases.Contains(cmdName, StringComparer.OrdinalIgnoreCase) &&
+                    commandAst.CommandElements.Any(
+                        e => e is CommandParameterAst parameterAst &&
+                                parameterAst.ParameterName.Equals("session", StringComparison.OrdinalIgnoreCase));
```
```suggestion
                if (!_invokeCommandCmdletNamesAndAliases.Contains(cmdName, StringComparer.OrdinalIgnoreCase))
                {
                    return false;
                }
                
                foreach (CommandElementAst commandElement in commandAst.CommandElements)
                {
                    if (commandElement is CommandParameterAst parameterAst
                        && parameterAst.ParameterName.Equals("session", StringComparison.OrdinalIgnoreCase))
                    {
                        return true;
                    }
                }
                
                return false;
````

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,493 @@
+                return _invokeCommandCmdletNamesAndAliases.Contains(cmdName, StringComparer.OrdinalIgnoreCase) &&
+                    commandAst.CommandElements.Any(
+                        e => e is CommandParameterAst parameterAst &&
+                                parameterAst.ParameterName.StartsWith("com", StringComparison.OrdinalIgnoreCase));
```
```suggestion
                if (!_invokeCommandCmdletNamesAndAliases.Contains(cmdName, StringComparer.OrdinalIgnoreCase))
                {
                    return false;
                }
                
                foreach (CommandElementAst commandElement in commandAst.CommandElements)
                {
                    if (commandElement is CommandParameterAst parameterAst
                        && parameterAst.ParameterName.StartsWith("com", StringComparison.OrdinalIgnoreCase))
                    {
                        return true;
                    }
                }
                
                return false;
```

rjmholt commented on Rules/UseUsingScopeModifierInNewRunspaces.cs:
```diff
@@ -0,0 +1,493 @@
+
+            private static readonly string[] s_dscScriptResourceCommandNames = {"GetScript", "TestScript", "SetScript"};
+
+            private readonly IEnumerable<string> _jobCmdletNamesAndAliases =
```
From the image you've shared above of test performance, it looks like I'm wrong about the perf hit here.

We should change these to `static readonly` though, since I suspect the reason the performance is alright is because the calls are cached. Meaning it's pointless for us to requery

Jawz84 commented on Rules/Strings.resx:
```diff
@@ -1,1125 +1,1140 @@
-    <value>The parameter '{0}' has been declared but not used. </value>	
-  </data>	
-  <data name="ReviewUnusedParameterName" xml:space="preserve">	
-    <value>ReviewUnusedPar
```
I'll look into it, I hadn't seen this, thanks.

Jawz84 replied:
it was the line endings. fixed in latest commit.

Jawz84 commented:
Really enjoying your feedback @rjmholt! I am learning a lot.

Jawz84 commented:
@rjmholt back over to you, all good here.

Jawz84 commented:
@bergmeister, @rjmholt thanks for all the reviews. Is there anything more I can do here? I am uncertain what the next step is. Is this blocked on review from @JamesWTruher? 
Thanks in advance for enlightening me 🙂

bergmeister commented:
I think having at least one reviewer from Microsoft is good enough, this is our unwritten rule, so I'll therefore merge. Congrats and thanks for your hard work
